### PR TITLE
Feature/middleware review

### PR DESCRIPTION
**Pull request by julianbermudez**
This pull request attends the comments the pull request #55.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/93/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/93/)

File: src/cpp/middleware/FastEntities.cpp [hunk 1]
```diff
@@ -267,7 +280,10 @@ void FastDataReader::onSubscriptionMatched(fastrtps::Subscriber*, fastrtps::rtps
 
 void FastDataReader::onNewDataMessage(fastrtps::Subscriber *)
 {
-    on_new_data_cb_();
+    if (nullptr != on_new_data_cb_)
+    {
+        on_new_data_cb_();
```
Comment (lemunozm) [inline src/cpp/middleware/FastEntities.cpp]: Only for curiosity, there is a pretty shortcut for this pattern: 
```c
on_new_data_cb && on_new_data_cb_()
```

File: src/cpp/participant/Participant.cpp [hunk 2]
```diff
@@ -18,34 +18,44 @@
...
-Participant::~Participant()
+Participant* Participant::create(
+        const dds::xrce::ObjectId& object_id,
+        const dds::xrce::OBJK_PARTICIPANT_Representation& representation,
+        Middleware& middleware)
 {
-    middleware_->delete_participant(get_raw_id());
-}
+    bool created_entity = false;
+    uint16_t raw_object_id = uint16_t((object_id[0] << 8) + object_id[1]);
```
Comment (lemunozm) [inline src/cpp/participant/Participant.cpp]: Like the previous comment, maybe exists a way to factorize this code into a static method of `XRCEObject` or something. Only the `XRCEObject` will know how to perform these conversions.

Comment (lemunozm) [reply src/cpp/participant/Participant.cpp]: Maybe this function could be called from the `ObjectIDHash::operator()` function too.

File: include/uxr/agent/datareader/DataReader.hpp [hunk 3]
```diff
@@ -77,36 +78,44 @@ class ReadTimeEvent
...
 
-    DataReader(DataReader&&)      = delete;
+    DataReader(DataReader&&) = delete;
     DataReader(const DataReader&) = delete;
     DataReader& operator=(DataReader&&) = delete;
     DataReader& operator=(const DataReader&) = delete;
 
-    bool init_middleware(const dds::xrce::DATAREADER_Representation& representation,
-            const ObjectContainer& root_objects);
+    void release(ObjectContainer&) override {}
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply include/uxr/agent/datareader/DataReader.hpp]: No, you don't need, in fact, it gives you a warning of an unused variable.

File: src/cpp/middleware/FastEntities.cpp (hunk 1 above)
Comment (julianbermudez) [reply src/cpp/middleware/FastEntities.cpp]: xD, I'm not sure but it will be nice.

File: src/cpp/subscriber/Subscriber.cpp [hunk 4]
```diff
@@ -49,7 +42,23 @@ bool Subscriber::init_middleware(const dds::xrce::OBJK_SUBSCRIBER_Representation
             break;
         }
     }
-    return rv;
+
+    return (created_entity ? new Subscriber(object_id, participant) : nullptr);
```
Comment (BorjaOuterelo) [inline src/cpp/subscriber/Subscriber.cpp]: We should avoid raw pointers. Maybe use smart pointers.

File: src/cpp/topic/Topic.cpp [hunk 5]
```diff
@@ -19,40 +19,49 @@
...
-            rv = middleware_->create_topic_from_xml(get_raw_id(), participant_->get_raw_id(), xml);
+            created_entity = middleware.create_topic_from_xml(raw_object_id, participant->get_raw_id(), xml);
             break;
         }
         default:
             break;
     }
-    return rv;
+
+    return (created_entity ? new Topic(object_id, participant) : nullptr);
```
Comment (BorjaOuterelo) [inline src/cpp/topic/Topic.cpp]: We should avoid raw pointers. Maybe use smart pointers.

File: src/cpp/datareader/DataReader.cpp [hunk 6]
```diff
@@ -26,88 +26,82 @@ namespace uxr {
...
             uint16_t topic_id;
-            if (middleware_->create_datareader_from_ref(
-                        get_raw_id(),
-                        subscriber_->get_raw_id(),
-                        ref,
-                        topic_id,
-                        std::bind(&DataReader::on_new_message, this)))
+            if (middleware.create_datareader_from_ref(raw_object_id, subscriber->get_raw_id(), ref, topic_id))
             {
                 dds::xrce::ObjectId topic_xrce_id = {uint8_t(topic_id >> 8), uint8_t(topic_id & 0xFF)};
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/datareader/DataReader.cpp]: Done! Thank you for the advice and help!

File: src/cpp/participant/Participant.cpp (hunk 2 above)
Comment (julianbermudez) [reply src/cpp/participant/Participant.cpp]: Done! Thank you for the advice and help!

File: src/cpp/middleware/FastEntities.cpp [hunk 7]
```diff
@@ -242,6 +243,18 @@ bool FastDataReader::match_from_xml(const std::string& xml)
     return rv;
 }
 
+bool FastDataReader::set_on_new_data_cb(OnNewData on_new_data_cb)
+{
+    on_new_data_cb_ = on_new_data_cb;
+    return true;
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/middleware/FastEntities.cpp]: I'm sure how the next centralized middleware is going to work.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/95/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/95/)

File: src/cpp/topic/Topic.cpp (hunk 5 above)
Comment (julianbermudez) [reply src/cpp/topic/Topic.cpp]: Done!

File: src/cpp/subscriber/Subscriber.cpp (hunk 4 above)
Comment (julianbermudez) [reply src/cpp/subscriber/Subscriber.cpp]: Done!

File: src/cpp/datawriter/DataWriter.cpp [hunk 8]
```diff
@@ -21,67 +21,66 @@
...
+                created_entity = true;
             }
             break;
         }
         default:
             break;
     }
-    return rv;
+
+    return (created_entity ? new DataWriter(object_id, publisher, topic) : nullptr);
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/datawriter/DataWriter.cpp]: Done!

File: src/cpp/datareader/DataReader.cpp [hunk 9]
```diff
@@ -26,88 +26,82 @@ namespace uxr {
...
+                created_entity = true;
             }
             break;
         }
         default:
             break;
     }
-    return rv;
+
+    return (created_entity ? new DataReader(object_id, subscriber, topic) : nullptr);
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/datareader/DataReader.cpp]: Done!

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/96/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/96/)

File: include/uxr/agent/datareader/DataReader.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (lemunozm) [reply include/uxr/agent/datareader/DataReader.hpp]: Ohh, ok, It is an inline declaration too

File: src/cpp/datareader/DataReader.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (lemunozm) [reply src/cpp/datareader/DataReader.cpp]: :D

File: include/uxr/agent/middleware/Middleware.hpp [hunk 10]
```diff
@@ -75,23 +73,25 @@ class Middleware
...
+    virtual bool matched_topic_from_xml(uint16_t topic_id, const std::string& xml) const = 0;
 
-    virtual bool matched_datawriter_from_ref(uint16_t datawriter_id, const std::string& ref) = 0;
-    virtual bool matched_datawriter_from_xml(uint16_t datawriter_id, const std::string& xml) = 0;
+    virtual bool matched_datawriter_from_ref(uint16_t datawriter_id, const std::string& ref) const = 0;
+    virtual bool matched_datawriter_from_xml(uint16_t datawriter_id, const std::string& xml) const = 0;
 
-    virtual bool matched_datareader_from_ref(uint16_t datawriter_id, const std::string& ref) = 0;
-    virtual bool matched_datareader_from_xml(uint16_t datawriter_id, const std::string& xml) = 0;
+    virtual bool matched_datareader_from_ref(uint16_t datawriter_id, const std::string& ref) const = 0;
```
Comment (lemunozm) [inline include/uxr/agent/middleware/Middleware.hpp]: The variable name `datawriter_id` must be `datareader_id`

File: include/uxr/agent/middleware/Middleware.hpp [hunk 11]
```diff
@@ -75,23 +73,25 @@ class Middleware
...
+    virtual bool matched_datareader_from_xml(uint16_t datawriter_id, const std::string& xml) const = 0;
```
Comment (lemunozm) [inline include/uxr/agent/middleware/Middleware.hpp]: The variable name `datawriter_id` must be `datareader_id`

File: include/uxr/agent/middleware/Middleware.hpp (hunk 10 above)
Comment (julianbermudez) [reply include/uxr/agent/middleware/Middleware.hpp]: Fixed!

File: include/uxr/agent/middleware/Middleware.hpp (hunk 11 above)
Comment (julianbermudez) [reply include/uxr/agent/middleware/Middleware.hpp]: Fixed!

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/97/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/97/)

**Pull request merged**